wai-extra/Network/Wai/Middleware/Select.hs

```diff
@@ -0,0 +1,91 @@
+---------------------------------------------------------
+-- |
+-- Module        : Network.Wai.Middleware.Select
+-- Copyright     : Michael Snoyman
+-- License       : BSD3
+--
+-- Maintainer    : Michael Snoyman <michael@snoyman.com>
+-- Stability     : Unstable
+-- Portability   : portable
+--
+-- Dynamically choose between Middlewares
+--
+-- It's useful when you want some 'Middleware's applied selectively.
+--
+-- Example: do not log health check calls:
+--
+-- > import Network.Wai
+-- > import Network.Wai.Middleware.HealthCheckEndpoint
+-- > import Network.Wai.Middleware.RequestLogger
+-- >
+-- > app' :: Application
+-- > app' =
+-- >   selectMiddleware (MiddlewareSelection $ selectMiddlewareExceptRawPathInfo "/_healthz" logStdout)
+-- >     $ healthCheck app
+--
+-- @since 3.1.10
+--
+---------------------------------------------------------
+module Network.Wai.Middleware.Select
+  ( -- * Middleware selection
+    MiddlewareSelection (..),
+    selectMiddleware,
+
+    -- * Helpers
+    selectMiddlewareOn,
+    selectMiddlewareOnRawPathInfo,
+    selectMiddlewareExceptRawPathInfo,
+    passthroughMiddleware,
+  )
+where
+
+import Control.Applicative ((<|>))
+import Data.ByteString (ByteString)
+import Data.Maybe (fromMaybe)
+import Network.Wai
+
+--------------------------------------------------
+-- * Middleware selection
+--------------------------------------------------
+
+-- | Relevant Middleware for a given 'Request'.
+newtype MiddlewareSelection = MiddlewareSelection
+  { applySelectedMiddleware :: Request -> Maybe Middleware
+  }
+
+instance Semigroup MiddlewareSelection where
+  MiddlewareSelection f <> MiddlewareSelection g =
+    MiddlewareSelection $ \req -> f req <|> g req
+
+instance Monoid MiddlewareSelection where
+  mempty = MiddlewareSelection $ const $ Just passthroughMiddleware
```

Comment:
This doesn't look law-abiding to me. It should probably be `const Nothing`. I don't think it's just an issue of laws, I think the current implementation could lead to surprising behavior.

Reply:
Good catch, thanks